src/core/Akka/Serialization/ReferenceResolver.cs

```diff
@@ -0,0 +1,55 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="NewtonSoftJsonSerializer.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2021 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2021 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Runtime.Serialization;
+using Newtonsoft.Json.Serialization;
+
+namespace Akka.Serialization
+{
+
+    public partial class NewtonSoftJsonSerializer
+    {
+        internal class ReferenceResolver : IReferenceResolver
```

Comment:
I'm wondering whether adapting [This IRefrenceResolver fix](https://github.com/JamesNK/Newtonsoft.Json/pull/1393/files), or [this one](https://github.com/JamesNK/Newtonsoft.Json/pull/2170/files) would be better overall and help us avoid the threadlocal? My worry is the performance as well as cognitive overhead of `WeakReference` in this implementation.